packages/dds/tree/src/internalDocumentationIndex.ts

```diff
@@ -0,0 +1,31 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * This file is meant to ensure that all public members have corresponding documentation published.
+ * This includes code that is not normally package-exported.
+ */
+/* eslint-disable no-restricted-syntax */
+export * from "./core/internalDocumentationIndex";
+export * from "./domains";
+export * from "./feature-libraries";
+export * from "./shared-tree";
+export * from "./shared-tree-core";
+export {
```

Comment:
I'm not a huge fan of these files. Maintaining them is non trivial since they can have name collisions, and files can get added, and new folders can get added. Also I don't think the set of things that are file exported is really the set of things we want docs for. Often things are file exported just for testing purposes and it seems wrong to couple that docs. I'd like docs for everything ideally.

If we want the docs instead to cover all the public stuff, it will be pretty sad/annoying that some of the grouping of things we do when reexporting them (like with FieldKinds, TypedSchema and SchemaAware) won't show up in the docs in anything like their actual public form.



If we want to have docs which actually match the public APIs (for the things that are public) but also include all the internal things, I don't think unioning together everything into one flat namespace is a good solution.

I don't really like this either, but I think it might be better to do something more like:

```typescript
export * as core from "./core/internalDocumentationIndex";
export * as domains from "./domains/";
...
export * as Public from ".";
```

This ensures that the exported API for every module/folder ends up in the API docs under a "Public" entry, but you can also find the internal docs for every module inside that one by looking under its name.
This duplicates everything thats public, but also means the Public API is included (including the tweaks in their index files) alongside the internal ones, and they are scoped instead of flattened so the names can't collide.

Reply:
Oh wow, I forgot that JSDoc is actually a thing and not just what we call those comment types. Maybe we can use that 😄.

Reply:
Looking at https://api-extractor.com/pages/setup/generating_docs, looks like API documenter supports consuming a collection of json files for different things and stitching them together. So maybe running API-extractor on each directory would work with that.

There is also the higher level question of what problem we are trying to solve. If what we really want is a way to link between doc comments for things that are not exported, this already works in intelisense in VS code. If we want it to stop breaking the API-extractor run and/or want the to be able to link into the source from non-source docs there may be other solutions. For example, if I want to link code from our markdown docs in a way that works in vscode and github, I can just factor the thing I want to link into its own file, and link the file. Sure that takes you to the source not a nice rendered page, but personally I'm fine with that.

Reply:
That works, but it's definitely not an ideal solution. Aside from the nice rendering (which I probably care more about than most people), links from the source won't work in github so if you want to navigate between related concepts, there's some searching involved. We're already doing a lot of linking in the jsdoc comments and I think it makes sense to take advantage of that. I'm also not sure how stable those that kind of manual linking would be. At least with an automated process, things should stay relatively up to date or be easier to fix if they don't.

I think the main advantage of having the docs nicely published, though, will be that we can do some more sophisticated things with them. In particular, I think having generated mind maps and backlinks would make navigating all of our different concepts easier.

Reply:
I realized that we can probably already control the doc bloat since things that don't really need docs just won't have a doc comment on them.

I spoke to @Josmithr and we want to work on better support for generating docs for everything during the next hackathon. In the meantime though, I'll continue to work on general documentation improvements. I also think it'd be reasonable for this change to go in as just a temporary measure, with the understanding that we won't need to maintain these files for terribly long.

Reply:
if we are going to keep these files, add a link to instructions for how to maintain them to the doc comment above. that will give you a centralized place to explains the plan for the future, and how to update these as needed.